### PR TITLE
APM Tracing

### PR DESCRIPTION
**Pull request by rishabhsri18**
Added a new context tracing in the context provided to fetch the details required to generate the traces for the function.

### CODE REVIEW

**Comment by isaaclipszyc** [comment]
Looks good so far, the only thing I see missing is the unit tests. Have you updated them to test some of these changes?

File: fdk.go
Comment (reclaro) [inline fdk.go:151]: I agree with Isaac that that is not part of the TracingContext interface.
In addition to it a small nit: in golang is not common to use the Get for the Get methond, can you please get rid of the Get prefix in these methods definitions?

File: fdk.go
Comment (reclaro) [inline fdk.go:157]: Same as a above 

small nit: in golang is not common to use the Get for the Get methond, can you please get rid of the Get prefix in these methods definitions?

File: fdk.go
Comment (reclaro) [inline fdk.go:164]: small nit: in golang is not common to use the Get for the Get methond, can you please get rid of the Get prefix in these methods definitions?

File: fdk.go
Comment (reclaro) [inline fdk.go:171]: small nit: in golang is not common to use the Get for the Get methond, can you please get rid of the Get prefix in these methods definitions?

File: fdk.go
Comment (reclaro) [inline fdk.go:178]: small nit: in golang is not common to use the Get for the Get methond, can you please get rid of the Get prefix in these methods definitions?

File: fdk.go
Comment (reclaro) [inline fdk.go:185]: small nit: in golang is not common to use the Get for the Get methond, can you please get rid of the Get prefix in these methods definitions?

File: fdk.go
Comment (reclaro) [inline fdk.go:192]: small nit: in golang is not common to use the Get for the Get methond, can you please get rid of the Get prefix in these methods definitions?

File: fdk.go
Comment (reclaro) [inline fdk.go:206]: small nit: in golang is not common to use the Get for the Get methond, can you please get rid of the Get prefix in these methods definitions?

File: fdk.go
Comment (reclaro) [inline fdk.go:251]: I think we want to be sure that the serviceName is lowercase, please add it

File: handler.go
Comment (reclaro) [inline handler.go:182]: if that header is not there we will get back an error from the ParseBool calll that means we set up the tctx.sampled = true. 
I think that is correct, can you just confirm that that is the logic we want?

**Review by reclaro — CHANGES_REQUESTED** [review]
Overall looks good, some small changes and i agree with Isaac comments.
Please add some unit tests that cover the different potential configuration that we can have with the presence or not of the different new ENV variables.

File: fdk.go
Comment (templecloud) [inline fdk.go:152]: This is just called isEnabled in the NodeFDK. Dont know about the others. Just a nit.

File: handler.go
Comment (templecloud) [inline handler.go:174]: Does this support canonicalised headers too? The Node FDK one does.

File: fdk.go
Comment (reclaro) [inline fdk.go:126]: thanks for the changes, I think that ServiceName is part of the TracingContext as it has defined as part of that change. It could be confusing having a serviceName method here as users can expect different values. Inside the tracingContext, instead, it is more clear what it means.

**Review by reclaro — COMMENTED** [review]
Just a small change and Ithink we are good to go.

File: fdk.go
Comment (isaaclipszyc) [reply fdk.go:126]: Yea I agree with Andrea, service name should be in the tracingContext as it is specific for tracing.

File: fdk.go
Comment (isaaclipszyc) [reply fdk.go:152]: In the python FDK it's is_tracing_enabled

Comment (rishabhsri18) [reply fdk.go:148]: Even in java it is isTracingEnabled

File: handler.go
Comment (rishabhsri18) [reply handler.go:174]: Not sure but using the go http library

**Review by reclaro — COMMENTED** [review]
LGTM wait for Tim to see if he is happy about his question about headers.

File: fdk.go
Comment (templecloud) [reply fdk.go:148]: Node looses then! I guess that slipped through. Hard to get the different language interfaces exactly the same. The term trace seemed a bit redundant in the node context: so it is 'traceContext.isEnabled' not 'traceContext.isTraceEnabled'.

**Review by templecloud — COMMENTED** [review]
Do we want to squash all these commits before merging?

**Review by isaaclipszyc — APPROVED** [review]
Comments have been addressed, changes look good to me. But yes, I agree with Tim, squashing the commits first is a good idea

**Review by reclaro — APPROVED** [review]
Please squash commits

File: fdk.go
Comment (rishabhsri18) [reply fdk.go:148]: Confused should I do this change or not

**Comment by rishabhsri18** [comment]
Squashed the commits

**Pull request merged**
